exonum/src/messages/mod.rs

```diff
@@ -284,3 +285,36 @@ impl PartialEq<Signed<RawTransaction>> for SignedMessage {
         self.eq(other.signed_message())
     }
 }
+
+impl<'de, T> Deserialize<'de> for Signed<T>
+where
+    T: ProtocolMessage + Deserialize<'de>,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        struct SignedLayout<P> {
+            payload: P,
+            #[serde(with = "HexStringRepresentation")]
+            message: SignedMessage,
+        };
+
+        let inner: SignedLayout<T> = SignedLayout::deserialize(deserializer)?;
+        // Checks that signed message correspond to the payload.
+        let encoded_payload = inner
+            .payload
+            .encode()
```

Comment:
I think it would be better to check the signature on the payload instead of reencoding it. Is is possible?

I remember that one of the pros of the new message layout was that we don't have to do this weird back-and-forth where deserialization requires serialization and vice versa.